keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2636 @@
+in the first alpha release of this enhancement due to the size and complexity
+of this enhancement.
+
+#### Namespace scoped policy configuration
+
+For phase 1, policy configuration resources were only allowed to be cluster
+scoped. We can support namespace scoped policy configuration as follows:
+
+- If the configuration resource is namespace scoped, it implicitly matches
+  resources only in the namespace it is in, but may further constrain what
+  resources it matches with additional match criteria.
+
+Benefits: Allows policy of a namespace to be controlled from within the
+namespace. 
+
+#### CEL expression scoping
+
+<<[UNRESOLVED jpbetz, ?? ]>>
+This is not yet well defined or agreed on. Reach agreement or drop.
+<<[/UNRESOLVED]>>
+
+CRD validation rules are scoped to the schema at the location in the OpenAPIv3
+where they are defined. This make validation rules far easier to author by
+eliminating the need to dereference from the root of an object to the field that
+needs to be validated. Should we provide something similar?
+
+Use cases:
+
+- Dereferencing to deeply nested fields in CEL. A `scope` field containing a
+  field path be sufficient. E.g. `spec.containers[*].image`
+- More ambitiously, some way to scope a CEL expression to a type found nested in
+  multiple kinds, e.g. "io.k8s.api.core.v1.PodTemplateSpec" or
+  "io.k8s.api.core.v1.Container" would help policy authors apply policies more
+  broadly and uniformly (e.g. to match initContainers, containers,
+  ephemeralContainers). Note that a list of field scopes described below can
+  handle some cases like this quite well (although each place a type exists must
+  be explicitly listed).
+
+For example, to validate all containers:
+
+```yaml
+  validations:
+    - expression: "self.name.startsWith('xyz-')"
+      scopes: ["spec.containers[*]"]
+```
+
+This could be extended to validate all kinds of containers in a pod if scopes
+supports a homogeniously typed list:
+
+```yaml
+  validations:
+    - expression: "self.name.startsWith('xyz-')"
+      scopes: ["spec.containers[*]", "initContainers[*]", "spec.ephemeralContainers[*]"]
+```
+
+`ephemeralContainers` might not exist due to version skew (it's a relatively new
+field), in which case that scope would be dropped (but a warning would be added
+to the status).
+
+#### Secondary Authz
+
+<<[UNRESOLVED jpbetz, deads2k, tallclair, ?? ]>>
+We have general agreement to include this as a feature, but need to provide a
+more concrete design.
+<<[/UNRESOLVED]>>
+
+kube-apiserver authorizer checks (aka Secondary-authz checks) have been proposed
+as a way of doing things like:
+
+- Validate that only a user with a specific permission can set a particular
+  field.
+- Validate that only a controller responsible for a finalizer can remove it from
+  the finalizers field.
+
+This could be supported by matching criteria, or via CEL expression access, or both.
+ 
+Use cases:
```

Comment:
not to self, I still owe you an answer about whether we only check self or whether we choose identities based on a field.  I think I remember seeing identify chosen by field in the open-cluster-management enhancements repo.